extensions/amp-analytics/0.1/visibility-impl.js

```diff
@@ -159,27 +159,32 @@ export function getElement(ampdoc, selector, el, selectionMethod) {
     return null;
   }
 
+  const elWin = el.ownerDocument.defaultView;
   // Special case for root selector.
   if (selector == ':host' || selector == ':root') {
-    const elWin = ampdoc.win;
     const parentEl = elWin.frameElement && elWin.frameElement.parentElement;
     if (parentEl) {
       return closestBySelector(parentEl, '.-amp-element');
     }
   }
 
+  let foundEl;
   if (selectionMethod == 'closest') {
     // Only tag names are supported currently.
-    const closestEl = closestByTag(el, selector);
-    // Restrict result to be contained by ampdoc.
-    if (closestEl && ampdoc.contains(closestEl)) {
-      return closestEl;
-    }
-    return null;
+    foundEl = closestByTag(el, selector);
   } else if (selectionMethod == 'scope') {
-    return el.parentElement.querySelector(selector);
+    foundEl = el.parentElement.querySelector(selector);
   } else if (selector[0] == '#') {
-    return ampdoc.getElementById(selector.slice(1));
+    foundEl = el.ownerDocument.getElementById(selector.slice(1));
```

Comment:
here is the summary. We can discuss it more over chat or gvc:

- Analytics can be done on a page by using amp-analytics tag. There can be more than one of these tags. `el` in this function is one such amp-analytics tag.

- `el` contains some config that specifies that certain behavior for another element needs to be tracked. The element is identified using `selector` and `selectionMethod`. Lets call this element `trackedElement`.

- Generally, `trackedElement` should always be in the same document as the `el`.  So if `el` is in shadow dom or friendly iframe, `trackedElement` should be in the shadow dom or friendly iframe as well. 
  - One exception to this is when ':root' selector is specified. In this case, we travel through the ancestor chain and return the first `.-amp-element`. This allows for tracking of shadow-dom or iframe as a whole unit.

- How we search for `trackedElement` is determined by `selectionMethod`. 
  - for `selectionMethod == closest`, we only look in ancestor chain and find the element that matches.
  - for `selectionMethod == scope`, we only look for elements in parent element.
  - for no `selectionMethod`, we just search the DOM for the element.

hope this helps...